### ISSUE TITLE
Anchor failed to build with provided dockerfile

### ISSUE THREAD

**Issue by yi-fan-song** [coral-xyz/anchor#2589]
Trying to build the docker file [here](https://github.com/coral-xyz/anchor/blob/be8764b8a827c87b347b7af0906bdbf0bd921a48/docker/build/Dockerfile#L24) results in `error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory`.

solana-cli: v1.14.22
anchor: v0.28.0

This arised from trying to fix anchor build on ubuntu 22.04 with the same error. Upgrading solana-cli to 1.16.5 fixed it, but then all subsequent programs compiled and running on the local test validator end up with [this](https://github.com/solana-labs/solana/issues/30271) kind of issue

**Comment by acheroncrypto** [comment]
The image is uploaded at [backpackapp/build](https://hub.docker.com/r/backpackapp/build). We're also using Ubuntu latest [in our CI](https://github.com/coral-xyz/anchor/blob/2af9cc669b734ca4b3a1ef4f590eec2a2dc097e5/.github/workflows/reusable-tests.yaml#L37) with only [these](https://github.com/coral-xyz/anchor/blob/2af9cc669b734ca4b3a1ef4f590eec2a2dc097e5/.github/actions/setup/action.yaml#L6) build dependencies.

> Upgrading solana-cli to 1.16.5 fixed it

`libz.so.1: cannot open shared object file` is not related to your `solana-cli` version, it could be that you had a corrupted installation and upgrading fixed it somehow.

From [this](https://stackoverflow.com/questions/21256866/libz-so-1-cannot-open-shared-object-file) StackOverflow question:

```sh
sudo apt install zlib1g
```

should install the necessary build dependencies.

> all subsequent programs compiled and running on the local test validator end up with https://github.com/solana-labs/solana/issues/30271 kind of issue

The problem in https://github.com/solana-labs/solana/issues/31960 is not yet fixed and it's not an Anchor specific issue.

**Comment by yi-fan-song** [comment]
Oh my apologies, I pasted the wrong error it's libssl not libz that's giving the issue

```
251.2 /root/.rustup/toolchains/bpf/bin/cargo: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```

this is why I had suspicions that it had to do with the newer ubuntu versions as 22.04 comes with libssl-dev3.0.9-1 while older versions of ubuntu had libssl-dev1.1.1. Of course I can manually downgrade the version of libssl-dev, but I would prefer not to resort to doing that.
Nonetheless, I think this is actually caused by cargo-build-bpf. Correct me if I'm wrong, is anchor running cargo-build-bpf and would it be possible for anchor to use cargo-build-spf instead as cargo-build-bpf has been deprecated. cargo-build-spf supposedly has that fixed from https://github.com/solana-labs/solana/issues/31128, but there is still https://github.com/solana-labs/solana/issues/31693. I would test it right now, but I don't have the machine on ubuntu 22.04.

Thanks for pointing me to https://github.com/solana-labs/solana/issues/31960

**Comment by acheroncrypto** [comment]
> Nonetheless, I think this is actually caused by cargo-build-bpf. Correct me if I'm wrong, is anchor running cargo-build-bpf and would it be possible for anchor to use cargo-build-spf instead as cargo-build-bpf has been deprecated.

You can use `anchor build --arch sbf` to build with `cargo-build-sbf` but the problem is that fix requires the newer version of build tools which has the https://github.com/solana-labs/solana/issues/31960 problem.

You can use [this](https://gist.github.com/siong1987/111937072c25cc5ec9f358e023dba63d) Dockerfile to build with `solana 1.14`.

**Comment by KS3D** [comment]
I've resolved the issue by making the following changes to the Dockerfile:

```Dockerfile
# Previous lines
ARG SOLANA_CLI
ARG ANCHOR_CLI

# Updated lines
ARG SOLANA_CLI="v1.14.26"
ARG ANCHOR_CLI="v0.28.0"

# Added these lines to download and install libssl1.1
RUN wget http://nz2.archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2.19_amd64.deb
RUN dpkg -i libssl1.1_1.1.1f-1ubuntu2.19_amd64.deb

# After install anchor command
# Also, added the following line to fix the error related to rustc version
RUN solana-install init 1.16.3

**Comment by acheroncrypto** [comment]
This problem wasn't caused by Anchor to begin with but it should be fixed in `0.29.0` regardless since we've dropped the Solana `1.14` support.
